bootstore/tests/v0-fsm-proptest-learner.rs

```diff
@@ -0,0 +1,479 @@
+            } => {
+                assert_eq!(request_id, expected_request_id);
+                assert_eq!(to, peer_id);
+        });
+        assert!(iter.next().is_none());
+    }
+
+    fn deliver_share_responses(&mut self, envelopes: Vec<Envelope>) {
+        for envelope in envelopes {
+            let share = Share(
+                self.pkgs.get(&envelope.to).unwrap().common.share.clone(),
+            );
+            let request_id = envelope.msg.request_id();
+            let rsp =
+                Response { request_id, type_: ResponseType::Share(share) }
+                    .into();
+            let output = self.common.sut.handle_msg(
+                self.common.now,
+                envelope.to.clone(),
+                rsp,
+            );
+
+            // Do we have a `LoadRackSecret` request?
+            if let Some(test_req) =
+                self.common.load_rack_secret_requests.get_mut(&request_id)
+            {
+                test_req.acks.insert(envelope.to);
+                // We don't count the SUT, which has its own share
+                if test_req.acks.len() == self.common.threshold - 1 {
+                    assert_matches!(
+                        output,
+                        Ok(Some(ApiOutput::RackSecret { .. }))
+                    );
+                    self.common.load_rack_secret_requests.remove(&request_id);
+                } else {
+                    assert_matches!(output, Ok(None));
+                    assert!(self.common.sut.drain_envelopes().next().is_none());
+                }
+            } else {
+                // These are extra shares (after the threshold is reached)
+                assert_matches!(output, Ok(None));
+                assert!(self.common.sut.drain_envelopes().next().is_none());
+            }
+        }
+    }
+
+    fn check_connect_output(
+        &self,
+        peer_id: &Baseboard,
+        result: Result<(), ApiError>,
+        envelopes: &Vec<Envelope>,
+    ) {
+        assert!(result.is_ok());
+        for envelope in envelopes {
+            let request_id = envelope.msg.request_id();
+            assert_eq!(peer_id, &envelope.to);
+            assert_matches!(
+                &envelope.msg,
+                &Msg::Req(Request { type_: RequestType::GetShare { .. }, .. })
+            );
+            // An outstanding request must exist and the peer must
+            // not have acked for an envelope to be sent.
+            if let Some(test_req) =
+                self.common.load_rack_secret_requests.get(&request_id)
+            {
+                assert!(!test_req.acks.contains(&envelope.to));
+            } else {
+                panic!("Share sent without an outstanding load rack request");
+            }
+        }
+    }
+}
+
+proptest! {
+    #![proptest_config(
+        ProptestConfig {max_shrink_iters: 100000, ..ProptestConfig::default()})]
```

Comment:
Would recommend using the [`test-strategy`](https://docs.rs/test-strategy/latest/test_strategy/) crate here, means there isn't a `proptest!` declarative macro involved.